File: neetcode-150/Two_pointers/Container_with_most_water/solution.py

from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:

        max_area = 0

        l, r = 0, len(height) - 1

        while l < r:
            width = r - l
            min_height = min(height[l], height[r])
            
            max_area = max(max_area, width * min_height)

            if height[l] < height[r]:
                l += 1
            else:
                r -= 1

        return max_area
    
# test
height = [1,8,6,2,5,4,8,3,7]
sol = Solution()
print(sol.maxArea(height))

File: neetcode-150/Two_pointers/Container_with_most_water/solution.js
function maxArea(height){
    const sol = new Solution();
    return sol.max_area(height)
}

class Solution{
    max_area(height){
        let res = 0;

        let left = 0;
        let right = height.length - 1

        while(left < right){
            const width = right - left;
            const min_height = Math.min(height[left], height[right]);
            const current_area = width * min_height;

            res = Math.max(res, current_area);
            if(height[left] < height[right]){
                left++;
            } else{
                right--;
            }
        }

        return res;
    }
}